### ISSUE TITLE
This service provider doesn't support keyed services.

### ISSUE THREAD

**Issue by MikeAlhayek** [OrchardCMS/OrchardCore#14983]
### Describe the bug

It looks like we still don't fully support keyed service.

I tried to register services using implementation factory like this

```
services.AddKeyedScoped<OpenAIClient>("SomeKey", (sp, obj) =>
{
    var openAIOptions = sp.GetRequiredService<IOptions<OpenAIOptions >>().Value;

    if (string.IsNullOrEmpty(openAIOptions.Host))
    {
        throw new InvalidOperationException($"{nameof(openAIOptions.Host)} is not configured.");
    }

    if (string.IsNullOrEmpty(openAIOptions.ApiKey))
    {
        throw new InvalidOperationException($"{nameof(openAIOptions.ApiKey)} is not configured.");
    }

    return new OpenAIClient(new Uri(openAIOptions.Host), new AzureKeyCredential(openAIOptions.ApiKey), GetOpenAIClientOptions(openAIOptions.ApiVersion));
});
```

then in the controller, I tried to resolve the service like this 

```
public AdminController([FromKeyedServices("SomeKey")] OpenAIClient client)
{
    _client = client;
}
```
That gave an the following exception 

```
InvalidOperationException: This service provider doesn't support keyed services.
```

### Expected behavior
I expected that the controller will be instantiated with an instance of `OpenAIClient` what matched "SomeKey" key.

**Comment by hishamco** [comment]
@MikeAlhayek do you plan to submit a PR or shall I do it? coz I already faced one issue in the past

**Comment by MikeAlhayek** [comment]
@hishamco I have too many things going on that this moment. If you have the time, a PR form you would be amazing :)

**Comment by hishamco** [comment]
@MikeAlhayek I just finished a unit test to ensure that your case has been worked fine

**Comment by hishamco** [comment]
Here is the full unit test code

```csharp
[Fact]
public void GetKeyedService_RegisteredUsingImplementationFactory()
{
    // Arrange
    var services = new ServiceCollection();

    services.Configure<OpenAIOptions>(options =>
    {
        options.ApiKey = "123";
        options.ApiVersion = "2020-10-10";
        options.Host = "https://api.openai.com";
    });

    services.AddKeyedScoped("SomeKey", (sp, obj) =>
    {
        var openAIOptions = sp.GetRequiredService<IOptions<OpenAIOptions>>().Value;

        if (string.IsNullOrEmpty(openAIOptions.Host))
        {
            throw new InvalidOperationException($"{nameof(openAIOptions.Host)} is not configured.");
        }

        if (string.IsNullOrEmpty(openAIOptions.ApiKey))
        {
            throw new InvalidOperationException($"{nameof(openAIOptions.ApiKey)} is not configured.");
        }

        return new OpenAIClient(new Uri(openAIOptions.Host), new AzureKeyCredential(openAIOptions.ApiKey), GetOpenAIClientOptions(openAIOptions.ApiVersion));
    });

    services.AddScoped<AdminController>();

    var serviceProvider = services.BuildServiceProvider();

    // Act
    var controller = serviceProvider.GetRequiredService<AdminController>();

    // Assert
    Assert.IsType<OpenAIClient>(controller.Client);
}

private static OpenAIOptions GetOpenAIClientOptions(string apiVersion) => new();

class OpenAIClient
{
    public OpenAIClient(Uri uri, AzureKeyCredential credential, OpenAIOptions options)
    {
        
    }
}

class OpenAIOptions
{
    public string ApiKey { get; set; }

    public string ApiVersion { get; set; }

    public string Host { get; set; }
}

class AdminController([FromKeyedServices("SomeKey")] KeyedServiceTests.OpenAIClient client)
{
    public OpenAIClient Client { get; } = client;
}
```

Please make it fail if I miss something in your case

**Comment by hishamco** [comment]
If no action needed please close

**Comment by MikeAlhayek** [comment]
@hishamco the reason why your test case passed is because you are testing a standard asp.net core `ServiceCollection` which supports keyed services. In OC, we use [ShellScopeServices](https://github.com/OrchardCMS/OrchardCore/blob/1269f09bf1a4f2aa5ab823ff261d033cdf4b64cb/src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScopeServices.cs#L5) which currently does not support keyed services.

The good news is that I fixed this issue using PR #14998

**Comment by MikeAlhayek** [comment]
@hishamco not sure either. I kinds feel useless to me. But I am sure it was added for a valid reason. If you care, try removing it and see if everything works fine including background processes. It's worth looking into it.
